crates/lox-time/src/utc.rs

```diff
@@ -1,264 +1,179 @@
-use std::fmt::Display;
+use std::fmt::{self, Display, Formatter};
 
+use num::ToPrimitive;
 use thiserror::Error;
 
-use crate::base_time::BaseTime;
-use crate::calendar_dates::{CalendarDate, Date};
-use crate::errors::LoxTimeError;
-use crate::julian_dates::{Epoch, JulianDate, Unit};
-use crate::subsecond::Subsecond;
-use crate::wall_clock::WallClock;
+use crate::calendar_dates::{CalendarDate, Date, DateError};
+use crate::constants::i64::SECONDS_PER_HALF_DAY;
+use crate::deltas::TimeDelta;
+use crate::julian_dates::JulianDate;
+use crate::time_of_day::{CivilTime, TimeOfDay, TimeOfDayError};
 
 pub mod transformations;
 
-/// A UTC timestamp with additional support for fractional seconds represented with femtosecond
-/// precision.
-///
-/// The `UTC` struct provides the ability to represent leap seconds by setting the `second`
-/// component to 60. However, it has no awareness of whether a user-specified leap second is valid.
-/// It is intended strictly as an IO time format which must be converted to a continuous time format
-/// to be used in calculations.
+#[derive(Debug, Clone, Error, PartialEq, Eq, PartialOrd, Ord)]
+pub enum UtcError {
+    #[error(transparent)]
+    DateError(#[from] DateError),
+    #[error(transparent)]
+    TimeError(#[from] TimeOfDayError),
+    #[error("no leap second on {0}")]
+    NonLeapSecondDate(Date),
+    #[error("UTC is not defined for dates before 1960-01-01")]
+    UtcUndefined,
+}
+
 #[derive(Debug, Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
 pub struct Utc {
-    hour: u8,
-    minute: u8,
-    second: u8,
-    subsecond: Subsecond,
+    date: Date,
+    time: TimeOfDay,
 }
 
 impl Utc {
-    pub fn new(
-        hour: u8,
-        minute: u8,
-        second: u8,
-        subsecond: Subsecond,
-    ) -> Result<Self, LoxTimeError> {
-        if !(0..24).contains(&hour) || !(0..60).contains(&minute) || !(0..61).contains(&second) {
-            Err(LoxTimeError::InvalidTime(hour, minute, second))
-        } else {
-            Ok(Self {
-                hour,
-                minute,
-                second,
-                subsecond,
-            })
+    pub fn new(year: i64, month: u8, day: u8) -> Result<Self, UtcError> {
```

Comment:
But the builders still feel clunky to me and it might be better to get rid of them and have the `time!` and `utc!` declarative macros as the main high-level API.